File: python/구현/Prob1244.py

""" 실버 4. 스위치 켜고 끄기 """

sc = int(input())                                # 스위치 개수                 
state = [0] + list(map(int, input().split()))    # 각 스위치의 상태 (1 켜져있음 0 꺼져있음)   
student = int(input())                           # 학생 수                      

for num in range(student):
    ge, sw = map(int, input().split())      # 입력받으면서 스위치 변경
    s = 1

    if ge == 1:         # 남학생
        while sw * s < sc + 1:
            state[sw * s] = abs(state[sw * s] - 1)
            s += 1

    else:               # 여학생
        state[sw] = abs(state[sw] - 1)
        while sw - s > 0 and sw + s < sc + 1 and state[sw - s] == state[sw + s]:
            state[sw - s] = abs(state[sw - s] - 1)
            state[sw + s] = abs(state[sw + s] - 1)
            s += 1

for num in range(1, sc+1):
    print(state[num], end = " ")
    if num % 20 == 0:print()

File: 02 백준/02 구현/Java/src/Prob1244.java

/**
* title  : 1244. 스위치 켜고 끄기 (Silver 4)
* time   : ms
* memory : KB
*/

import java.io.*;
import java.util.*;

public class Prob1244 {
    
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());
        int[] state = new int[N+1];

        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 1; i <= N; i++) {
            state[i] = Integer.parseInt(st.nextToken());
        }

        int student = Integer.parseInt(br.readLine());
        for (int i = 0; i < student; i++) {

            st = new StringTokenizer(br.readLine());
            int gender = Integer.parseInt(st.nextToken());
            int number = Integer.parseInt(st.nextToken());

            if (gender == 1) {  // 남학생

                int pos = number;
                while (pos < N) {
                    state[pos] = Math.abs(state[pos] - 1);
                    pos += number;
                }
            } else {            // 여학생
                state[number] = Math.abs(state[number] - 1);

                int lpos = number;
                int rpos = number;
                while (0 <= --lpos && ++rpos < N && state[lpos] == state[rpos]) {
                    state[lpos] = Math.abs(state[lpos] - 1);
                    state[rpos] = Math.abs(state[rpos] - 1);
                }
            }
        }

        StringBuilder sb = new StringBuilder();
        for (int i = 1; i <= N; i++) {
            if (i % 20 == 0) {
                sb.append(state[i]).append("\n");        
            } else {
                sb.append(state[i]).append(" ");
            }
        }

        System.out.println(sb);
    }
}
